### ISSUE TITLE
Remove Assert.DoesNotThrow

### ISSUE THREAD

**Issue by bradwilson** [xunit/xunit#188]
So. Incredibly. Not. Needed.

**Comment by ploeh** [comment]
I, for one, liked the explicitness of being able to state that *this particular line of code shouldn't throw, and that's the test*.

Not that I use it a lot, but I do use it occasionally, so I don't think it should go.

I can certainly live without it, but I still think it adds value - not because it *does* anything, but because it makes it explicit what is being tested.

In the end, code is read more than it's written, and as Fowler writes in [Refactoring](http://amzn.to/1gDifab): *Any fool can write code that a computer can understand. Good programmers write code that humans can understand.*

I've always viewed the real value of `Assert.DoesNotThrow` to be making the test code more readable (because it's [more explicit](https://www.python.org/dev/peps/pep-0020)).

**Comment by bradwilson** [comment]
You're welcome to bring in the assertion library as source (NuGet package `xunit.assert.source`), which brings `Assert` as a partial class that you can add methods to, and add back `DoesNotThrow`.

**Comment by ploeh** [comment]
I don't think it'd compile :smiling_imp:

**Comment by bradwilson** [comment]
Why not?

**Comment by ploeh** [comment]
It's C# :wink:

**Comment by danielbaharestani** [comment]
While I agree with Mark, regarding explicitness of <code>DoesNotThrow</code>, I believe even if a method is going to be removed from a framework it's better to make it obsolete for a few versions rather than deleting it immediately.

**Comment by bradwilson** [comment]
This is a major new release, a complete re-write. To keep all functionality and mark it obsolete would've been impossible.

**Comment by ashmind** [comment]
Just got xunit 2.0 for the first time and this is a step back.

I agree with @ploeh -- this shouldn't have been removed. I actively used it to make it clear that assertion is not forgotten. I even had a helper method for that in dark ages of forced MSTest.

I've also used it to educate people on proper tests -- focused on one thing per test and not random set of lines where each can fail (yes, each can fail, but test should have a point).

**Comment by bradwilson** [comment]
I'm afraid we're at a philosophical point of difference.

> I actively used it to make it clear that assertion is not forgotten.

Over time, we've become convinced that a test which simply says "please don't throw" is almost never the right test. If there are no observable side-effects, how do you know any work was done?

**Comment by bradwilson** [comment]
Also, in case you haven't tried it, we have a source-based NuGet package for the Assert class, which is delivered as a multi-file partial class. That makes it trivial to bring in the existing assert library and then add whichever asserts you feel are missing.

**Comment by ashmind** [comment]
@bradwilson 
I do see your point, so to be more specific I checked where `DoesNotThrow` is used in my code.

Here are some cases:

1. Positive validation.
I only care that Save does not throw a `ValidationException` when fields are valid, but I don't care whether it's actually saved correctly (other tests check that).

2. Mass theory.
Consider a decompiler. I write a lot of detailed tests on specific case. But I can also make a Theory that feeds it all methods from a bunch of assemblies, just to check if there is some unexpected edge case. I can't give a reasonable expectation for each row in this Theory -- and "not throwing" is a good indicator to see which ones should be tested more.

3. "Does it work under those conditions"
This one is similar to the validation tests in that there are other tests for actual functionality, but you want to see if system does not fail under a certain condition. Some examples: does the action complete if MSDTC is disabled? does the action complete if it is run in parallel 10000 times? Some of them can check actual output, but sometimes you don't care much -- it should just not fail.

4. "More detailed test TBD"
Not that great from purist perspective, but sometimes you want to get a test in, and then improve on it. This test might be non-essential/nice-to-have, or the requirements/design might be in flux -- sometimes having a not-throws is more valuable than nothing at all.

> Also, in case you haven't tried it, we have a source-based NuGet package for the Assert class, which is delivered as a multi-file partial class. That makes it trivial to bring in the existing assert library and then add whichever asserts you feel are missing.

Sure, and I can add `AssertEx.DoesNotThrow` just as easily. However realistically I can't tell juniors on other projects "you just need to get an extra class that does nothing if you want your tests to be descriptive". :)

**Comment by ploeh** [comment]
> a test which simply says "please don't throw" is almost never the right test

Here's another counter-example: sometimes, I want to implement [`IObserver<T>`](https://msdn.microsoft.com/en-us/library/dd783449.aspx) (for some `T`), but often, I only need the System Under Test to implement the `OnNext` method.

For the two other methods, `OnError` and `OnCompleted`, I only want to test that they don't throw any exceptions when invoked.

I agree that in general, a test that only verifies that an action doesn't throw, is a test smell, but there are cases, as for example when implementing [Adapters](http://en.wikipedia.org/wiki/Adapter_pattern), when it's the correct thing to do.

This isn't something that I do much, and I can most likely live without `Assert.Throws`, but I certainly believe it has proper uses.

**Comment by JamesReate** [comment]
Another vote in favor of keeping Assert.DoesNotThrow

The only reason is code readability & understanding intent.

**Comment by SurajGupta** [comment]
another vote to bring this back.  Breaks backward compat

**Comment by adamralph** [comment]
Personally, I believe neither `Assert.Throws<T>` nor `Assert.DoesNotThrows<T>`  should exist:

```C#
// act
var ex = Record.Exception(() => sut.Foo());

// assert
Assert.IsType<ExpectedException>(ex);
```

or

```C#
// act
var ex = Record.Exception(() => sut.Foo());

// assert
Assert.Null(ex);
```

**Comment by ploeh** [comment]
@adamralph That's a nice idea! :+1: 

... although obviously, it should be based on a sane model of *there/not there*, like a [Maybe](https://hackage.haskell.org/package/base-4.7.0.2/docs/Data-Maybe.html)/[option](https://msdn.microsoft.com/en-us/library/dd233245.aspx), instead of null :trollface:

**Comment by adamralph** [comment]
`Record.Exception()` exists in xunit today.

**Comment by wholroyd** [comment]
@adamralph beat me to it

https://github.com/xunit/xunit/blob/master/src/xunit.assert/Asserts/Record.cs

**Comment by adamralph** [comment]
Not just 2.0 either, it's been in 1.x for a very long time.

**Comment by SurajGupta** [comment]
There are situations where frameworks should be highly opinionated and situations where frameworks are overly opinionated.  For example, I think xUnit's departures from NUnit were fantastic, [opinionated choices].  This paritcular issues, however, seems to be overly opinionated.

But what is the utility of getting rid of Throws/DoesNotThrow?  The arguments feel purist to me.  Oh it's so elegant!  Oh my code will be so much more beautiful and clean and clever!  It's a slippery slope towards "you should test exactly the way I test".  Throws/DoesNotThrow is a type of assertion.  The goal of any good testing framework is to make it easy and simple to express a wide variety of assertions.  It's perfectly fine to be opinionated about the names of the assertions (i.e. DoesNotThrow, ItDoesNotThrow, WillNotThrow), but to remove an assertion that many people find valuable, particularly one that's dead-simple to understand and is highly expressive seems ridiculous to me.

[opinionated choices]: https://xunit.codeplex.com/wikipage?title=WhyDidWeBuildXunit

**Comment by adamralph** [comment]
As you say, these are opinions. The above is just my opinion. What the framework owners take from that is their business.

**Comment by bradwilson** [comment]
I actually agree that, in retrospect, Assert.Throws wasn't a great idea. Fortunately, the right idea is also available to you without cost. :)

**Comment by GuardRex** [comment]
@adamralph Shout out ... Thanks! :+1:

I wish `Assert.DoesNotThrow` had remained, but your tip got me where I need to be.

**Comment by adamralph** [comment]
@GuardRex you're welcome :smile: 

@bradwilson have you considered obsoleting `Assert.Throws`?

**Comment by bradwilson** [comment]
I'll talk to Jim about it.

The difference between obsoleting `Assert.DoesNotThrow` and `Assert.Throws` comes down to differences in philosophy. The former literally did nothing, and it's hard to argue that it should've ever been included. At least the latter has value in brevity for developers who don't stick to strict AAA.

I think if we had it to do over again, we wouldn't have added it, but that's not the situation we're in right now. ;)

**Comment by yannduran** [comment]
@bradwilson what would you use instead? I'm brand new to xUnit, & have just got my head around the way xUnit does [ExpectedException]. I wanted to then then add a test to check opposite of my first assertion. I saw how @adamralph would handle the scenarios if those ,methods didn't exist. But what would you use if there's no Assert.Throws & Assert.DoesNotThrow

**Comment by bradwilson** [comment]
I would use `Record.Exception`.

If you're following AAA (Arrange/Act/Assert), then `Assert.Throws` violates the pattern by being both Act and Assert. So this:

```csharp
// Act & Assert
Assert.Throws<ArgumentException>(() => ...);
```

Becomes this:

```csharp
// Act
var ex = Record.Exception(() => ...);

// Assert
Assert.IsType<ArgumentException>(ex);
```

Note that `Record.Exception` returns `null` when the code in question did not throw.

**Comment by yannduran** [comment]
OK, perfect! Using **Assert.ThrowsException()** did actually feel uncomfortable, but I thought that was just because I was trying to use a new testing framework. 

Although I liked the syntax of the **Assert.ThrowsException** assertion itself, I didn't really like the way you have to use it in tests. When I googled "expected exceptions in xUnit", **Assert.ThrowsException()** is what I found. 

I had seen that Adam suggested using **Record.Exception()**, but that felt a bit too strange & I didn't look into it any further, & it didn't "click" until you pointed out the AAA violation explicitly. Looking back now, it all makes (& made) perfect sense.

Thanks for the explanation. I'll be using **Record.Exception()** myself from now on.

**Comment by Valeriy1991** [comment]
@bradwilson Nothing prevents you from doing so:
```csharp
// Act
Action act = () => ...;
//Assert
Assert.Throws<ArgumentException>(act);
```
instead of your example:
```csharp
// Act & Assert
Assert.Throws<ArgumentException>(() => ...);
```
In my opinion, the `Assert.DoesNotThrow` more readable and understandale than `Record.Exception`...

**Comment by adamralph** [comment]
@Valeriy1991 that doesn't really acheive anything, since the "act" is just assigning a variable. The act and assert are still conflated. It's better if the "act" has happened by the time you step down the "assert" line.

**Comment by ashmind** [comment]
Not a fan of 
```
var exception = Record.Exception(() => act());
Assert.Null(exception);
```

Feels weird to record exception that shouldn't even be there.

But on act vs assert, maybe it should be `Assert.DidNotThrow()`, even simpler -- no delegates.

**Comment by maiconheck** [comment]
Guys, I agree with both the arguments pro and against the `Throw` / `DoesNotThrows` methods. They are two ways to solve the problem.

But, as many of you have said, `Throw` / `DoesNotThrows` is much clearer and more intuitive. In addition, several developers that coming from NUnit are already familiar with this Api.

So, since this vote has already been closed, and `DoesNotThrows` will not be added nor the `Throw` removed, I have chosen to manually resolve the problem in my code:
If someone else is interested in doing the same:

```C#
/// <summary>
/// Provides an interface to base unit test classes in order to abstract
/// the different implementations between xUnit and NUnit and so on.
/// </summary>
public interface ITest
{
    /// <summary>
    /// Verifies that the exact exception is thrown.
    /// </summary>
    /// <param name="testCode">The test code.</param>
    /// <typeparam name="T">The exception type.</typeparam>
    void Throws<T>(in Action testCode) where T : Exception;

    /// <summary>
    /// Verifies that no exception was thrown.
    /// </summary>
    /// <param name="testCode">The test code.</param>
    void DoesNotThrows(Action testCode);
}
```

```C#
public abstract class XUnitTestBase : ITest
{
    public void Throws<T>(in Action testCode)
    	where T : Exception
    {
    	var ex = Record.Exception(testCode);
    	Assert.IsType<T>(ex);
    }
    
    public void DoesNotThrows(Action testCode)
    {
    	var ex = Record.Exception(testCode);
    	Assert.Null(ex);
    }
}
```

Then, I inherit from the `XUnitTestBase` in my test classes, in order to have the` Throws` / `DoesNotThrows` methods available.

I could have another implementation of `ITest` as` NUnitTestBase`, so I could isolate these differences between the Unit Test Apis. This is especially useful for legacy projects that use more than one Unit Test Framework, or are migrating from one to another.

**Comment by M-Yankov** [comment]
Is it OK to inherit `Assert` class? Like:

```c#
public class MyAssert : Assert
{
    public static void DoesNotThrow(Func<object> func)
    {
        func();
    }

    public static void DoesNotThrow(Action action)
    {
        action();
    }
}

```

**Comment by bradwilson** [comment]
Yes. That's why we didn't make it static.

**Comment by vers-one** [comment]
If you have code like this:
```csharp
public class MyClass
{
    private readonly string param;

    public MyClass(string param)
    {
        this.param = param ?? throw new ArgumentNullException(nameof(param));
    }
}
```

you would want to test that the constructor throws an `ArgumentNullException` when the `param` is `null`:
```csharp
[Fact(DisplayName = "MyClass constructor should throw ArgumentNullException if the param is null")]
public void ConstructorWithNullParamTest()
{
    Assert.Throws<ArgumentNullException>(() => new MyClass(null));
}
```

But before doing that it would be nice to establish a base line to make sure that the constructor of `MyClass` doesn't throw `ArgumentNullException` regardless of its input:
```csharp
[Fact(DisplayName = "Constructing MyClass with non-null param should succeed")]
public void ConstructorWithValidParamTest()
{
    new MyClass("test");
}
```

However, if you do that, you will get a [CA1806](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1806) warning from the compiler with the following message:
> ConstructorWithValidParamTest creates a new instance of MyClass which is never used. Pass the instance as an argument to another method, assign the instance to a variable, or remove the object creation if it is unnecessary.

There are workarounds to avoid this warning: you can use `Assert.Null(Record.Exception(() => new MyClass("test")))` or make your own `Assert` subclass with `DoesNotThrow()` implementation but it would be much easier if XUnit had `Assert.DoesNotThrow` out of the box.

**Comment by jnm2** [comment]
`_ = new MyClass("test");` also strikes me as fairly clearly testing that it doesn't throw.

**Comment by vers-one** [comment]
> `_ = new MyClass("test");`

Oh, yeah, that works too.

**Comment by shoter** [comment]
It is a pity it was removed. Now to have smoke tests with statis code analysis provided by SonarCloud we need to do out of the box Shenningans to make it work.

**Comment by bradwilson** [comment]
@shoter I have no idea what that means, but... just throw exceptions. Or write your own.
